lib/core/versions/latest/ThroughputLimiter.ts

```diff
@@ -0,0 +1,90 @@
+import AnchoredDataSerializer from './AnchoredDataSerializer';
+import IThroughputLimiter from '../../interfaces/IThroughputLimiter';
+import ITransactionStore from '../../interfaces/ITransactionStore';
+import PriorityQueue from 'priorityqueue';
+import TransactionModel from '../../../common/models/TransactionModel';
+import { SidetreeError } from '../../Error';
+import ErrorCode from './ErrorCode';
+
+/**
+ * rate limits how many operations is valid per block
+ */
+export default class ThroughputLimiter implements IThroughputLimiter {
+  private transactionsPriorityQueue: any;
+  public constructor (
+    private maxNumberOfTransactionsPerBlock: number,
+    private maxNumberOfOperationsPerBlock: number,
+    private transactionStore: ITransactionStore
+  ) {
+
+    const comparator = (a: TransactionModel, b: TransactionModel) => {
+      // higher fee comes first. If fees are the same, earlier transaction comes first
+      return a.transactionFeePaid - b.transactionFeePaid || b.transactionNumber - a.transactionNumber;
+    };
+
+    this.transactionsPriorityQueue = new PriorityQueue({ comparator });
+  }
+
+  /**
+   * Returns an array of transactions that should be processed. Ranked by highest fee paid per transaction and up to the
+   * max number of operations per block
+   * @param orderedTransactions The transactions that should be ranked and considered to process
+   */
+  public async selectQualifiedTransactions (orderedTransactions: TransactionModel[]): Promise<TransactionModel[]> {
+    if (!orderedTransactions.length) {
+      return [];
+    }
+
+    const currentBlockHeight = orderedTransactions[0].transactionTime;
+    for (const transaction of orderedTransactions) {
+      if (transaction.transactionTime !== currentBlockHeight) {
+        this.transactionsPriorityQueue.clear();
+        throw new SidetreeError(ErrorCode.TransactionsNotInSameBlock, 'transaction must be in the same block to perform rate limiting');
+      }
```

Comment:
You're right. I added it just in case we make a mistake in observer, we will know from this error. Also I think it doesn't hurt to be safe. What do you think?

Reply:
I don't feel too strongly about it, so we can keep if you'd like. I agree with defensive programming in principle, but in this context, there are many "checks" that we could have done in the current codebase but have not, so I ask myself what makes this check special and others not?

Reply:
I feel like the check should be there because the class and function shouldn't know how it's being used, so it shouldn't assume that the array is ordered. I will leave it for now. It's a quick change to remove it if you think otherwise.